pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/properties/PreserveStashesJobProperty.java

```diff
@@ -112,8 +126,8 @@ public void onCompleted(Run<?,?> r, TaskListener listener) {
                 WorkflowJob j = ((WorkflowRun) r).getParent();
                 PreserveStashesJobProperty prop = j.getProperty(PreserveStashesJobProperty.class);
                 if (prop != null && prop.getBuildCount() > 0) {
-                    // TODO: Determine if this is overzealous. Might need to ape log rotation more closely.
-                    for (WorkflowRun build : j.getBuilds().completedOnly()) {
+                    // The "+1" is to ensure that we look at one more than the maximum possible number of builds with saved stashes.
+                    for (WorkflowRun build : j.getBuilds().completedOnly().limit(MAX_SAVED_STASHES + 1)) {
```

Comment:
I guess you never managed to fix the test to fail if you omit `limit`?

Will probably neglect to clear stashes under some conditions, such as if you start off with a high limit and then edit `properties` to specify a lower limit.